### ISSUE TITLE
Inconvenient volume scaling for Alsa output

### ISSUE THREAD

**Issue by rschroll** [ejurgensen/forked-daapd#1079]
I've installed v 27.2 of forked-daapd on a Raspberry Pi 3 running Raspbian buster.  The volume scale for Alsa output (via the headphone jack) is such that only about 30% of the scale is useful.  If it can't be set automatically, it'd be nice to have some option to adjust how the scale is interpreted, so I don't have to futz with small adjustments to the slider.

It appears that, on this system at least, inputs to `snd_mixer_selem_set_playback_volume_all` are treated as being in units of hundredths of a decibel (a millibel, I guess).  The available range spans -102 dB to +4 dB.  This is an impressive, but practically useless, dynamic range.  I don't need to adjust values over that full range, and having that full range makes more practical adjustments more difficult.

A simple way to fix this would be to allow the user to specify a min volume that corresponded to 0% (or maybe 1%).  If I could set the bottom of the scale to -30 or -40 dB, I'd still have plenty of range, but I could actually have a use for most of it.

I feel a better, if more involved, solution would be to use a mapping that reflected perceived loudness, rather than dB.  The current behavior seems to reflect the behavior of `amixer -R`.  But there's another mode, `amixer -M`, that uses a "mapped volume" percentage to reflect our experience.  In my testing, the `-M` option feels much better that `-R`.  The range at the top of the scale is expanded and the lower end of the scale, where nothing is audible anyway, is compressed.  The algorithm is mentioned here: https://mailman.alsa-project.org/pipermail/alsa-devel/2012-March/050074.html.

This second method seems to be what other Alsa clients do.  For example, here is MPD converting volume percentage to dB levels: https://github.com/MusicPlayerDaemon/MPD/blob/fd71514068be53a98c8f4b87d3164fd632ab24cb/src/mixer/plugins/volume_mapping.c#L81. It seems to follow the procedure outlined in that email.

This may be the underlying issue in #564.  There it was suggested that this was a problem with Alsa, but I think Alsa is operating as specified.  It's specifying percentages in dB, even though that doesn't match how hearing works.  It's annoying that clients have to worry about it, but at least there's example code to crib off of.

Please let me know if there's any Alsa debug info I could provide that would help out.

**Comment by ejurgensen** [comment]
First of all, thanks for a great analysis of the issue.

I'm all aboard doing what other Alsa clients do, but the MPD approach seems pretty complicated. Do you think that is a standard approach?

I think it is suboptimal to have this type of sound configuration on an application level, it should be on a system level (via a global configuration). But I don't know if that is possible with Alsa - have you looked into that? If there just isn't a better way then I agree that we need to adapt forked-daapd.

I don't use Alsa much myself, so I would also like to ask for @whatdoineed2do's opinion on this.

**Comment by whatdoineed2do** [comment]
The code listed for mpd above is pretty much from [alsa-utils's](https://github.com/alsa-project/alsa-utils/) [alsamixer itself](https://github.com/alsa-project/alsa-utils/blob/master/alsamixer/volume_mapping.c).

I've not seen anything in the alsa docs that would request alsa subsystem to use 'perceived' volume or dB vol at a system scope for API calls -  however, since we would (in many cases - RPi - traditionally) use `alsamixer` to set the volume it would not be a bad thing to minic this behaviour.

The change is quite small (introduce `volume_mapping.c` from `alsa utils` and call their _set volume_ function instead of interpreting ourselves in https://github.com/ejurgensen/forked-daapd/blob/master/src/outputs/alsa.c#L292-L320).


I also to take a quick look at some other appls that have volume control and `mpv` takes a slightly different and more simplistic approach:

https://github.com/mpv-player/mpv/blob/68ade4e5a5e760bb5a4b793ebfa98ca386786605/audio/out/ao_alsa.c#L166-L180

which we can implement as a one line replacement but i havent been able to test this properly.
https://github.com/ejurgensen/forked-daapd/blob/master/src/outputs/alsa.c#L312
```
pcm_vol = volume / (100/(float)(mixer->vol_max - mixer->vol_min)) + mixer->vol_min + 0.5; 
```

This desktop I'm using is using pulseaudio as a backend (providing an alsa sink to forked daap) - i'll try to have a look at this later on a box that i've stripped pulse and only has alsa enabled for sound and see if the mpv approach works as well as the alsamixer.

**Comment by whatdoineed2do** [comment]
I tested this on this laptop where i've removed pulseaudio - only the alsamixer volume adjust algorithm does what we want.  Created a PR that uses a stripped down and modified alsamixer implmentation.

**Comment by rschroll** [comment]
Thanks for the quick reaction!  I'm no Alsa expert, but flipping through
the API docs didn't seem to reveal any better way to do this.  The fact
that everyone else is jumping through these same hoops suggests this is
what you have to do.

Unfortunately, I'm not set up to build forked-daapd for testing.  If it
would help, I can try to get set up.  But it sounds like it's already well
in hand.

**Comment by whatdoineed2do** [comment]
If can you try that would be useful.  This is the build from my PRI 4 running buster for testing.  No mpd enabled, just alsa.
[forked-daapd.gz](https://github.com/ejurgensen/forked-daapd/files/5089269/forked-daapd.gz)

**Comment by ejurgensen** [comment]
Thanks for the quick PR. I haven’t tested it myself, but one thing I like about it is that it doesn’t require user configuration.

@whatdoineed2do some PR comments: Can you integrate the mapping into alsa.c instead of having a new separate file? I hope we won’t need these shenanigans for other outputs, and if we do, we can always move it out. I also don’t get the -lm change, since math is already used other places in forked-daapd. Can that be removed?

@rschroll also thanks for checking if alsa offers a solution for this. I also looked a bit around, and I think you are right that there is none.

**Comment by whatdoineed2do** [comment]
Yes i can pull the `alsamixer` code into `alsa.c` - i'll drop out the 2 element array mappings;  i kept this in there mostly so we could easily diff against alsamixer's code if we needed (bug fixes etc - but given this file has had 4 upd since 2010 maybe we'll be good).  Two Qs:
* Regarding the copyright, just add `(c) 2010 Clemens Ladish ....` to the top of the `alsa.c` file and mark the block of code that we're bringing across?  Anything else special required?  Does this cause any issues?
* coding style - keep existing or use forked-daapd's?

Regarding `-lm`, on my fedora 32 desktop/linux, it required the explicit `-lm` at link stage otherwise linking failed - certainly looking at the generated `src/Makefile` I didnt see any reference to `-lm` but I'm also building without a bunch of options (`./configure --prefix=/usr --sysconfdir=/etc --disable-spotify --disable-lastfm --disable-chromecast --localstatedir=/var --libdir=/usr/lib/arm-linux-gnueabihf --without-pulseaudio --disable-mpd`)

**Comment by ejurgensen** [comment]
I've tested the PR on a Ubuntu vm, and it worked fine. It didn't really change the volume slider much, but that was a good thing because it was already sensible. I will also give it a go on one of my RPis, but I don't remember if the slider was incovenient on them. Before merging, it would be good to get a confirmation that the slider is actually improved on your device, @rschroll.

**Comment by whatdoineed2do** [comment]
This is more noticable when you try to set the volume to lower levels because the range at the lower end of the volume slider is "more compressed"

When we adjust the volume using `alsamixer` we can see the vertical level come down in increments that match our perception.   So 50% on `alsamixer` is probably 1/2 as loud as max volume.

On the existing `master` running on a RPi1 and headphone output we can observe `alsamixer` in a terminal and changing the slider on web ui to 1/2 way.  Again you'd expect this to be 1/2 as loud as full volume buts its significantly quieter - `alsamixer` shows the volume at 11%

With the PR, the volume adjusted to 1/2 way on the web ui slider the volume sounds right and `alsamixer` reports the volume at around 50% - this we'd expect since we're now using the same code from `alsamixer` to change the volume : )

But yes, @rschroll should test and confirm (i incl a binary further up in the thread) but it looks and sounds right.
